### ISSUE TITLE
Rule Change: Add option to func-names to only check async functions

### ISSUE THREAD

**Issue by tkalliom** [eslint/eslint#16182]
### What rule do you want to change?

func-names

### What change to do you want to make?

Generate fewer warnings

### How do you think the change should be implemented?

A new option

### Example code

```js
function doUpdate(records) {
  const [ids, updates] = records.reduce(
    function (acc, currentRecord) {
      const { id, ...attributes } = currentRecord;
      acc[0].push(id);
      acc[1].push({ id, attributes });
    },
    [[], []]
  )

  db.transaction(
    async function (trx) {
      const count = trx.findByIds(ids);
      if (count < records.length)
        throw new Error("Invalid id reference");

      trx.update(updates);
    }
  );
}
```

### What does the rule currently do for this code?

Two errors are output: `Unexpected unnamed function` for the first callback, and `Unexpected unnamed async function` for the second one.

### What will the rule do after it's changed?

With a new option, e.g. `sync: false` or `ignoreSync: true`, an error would be produced only for the second callback function.

### Participation

- [ ] I am willing to submit a pull request to implement this change.

### Additional comments

Not allowing any anonymous functions at all is, IMHO, a bit tedious. Callbacks can often be so concise that it would feel contrived trying to come up with a name for them. And if an error is produced in one, the stack trace usually remains moderately understandable – in the example code, two levels above the callback is the named `doUpdate`.

When the function is `async`, the need for a name becomes greater. If an error is thrown, only runtime internals (like `processTicksAndRejections` and `runMicrotasks`) are above the `async` function in the stack trace. (The way we ran into this was Sentry grouping together issues where an error was thrown from an `async` callback).

**Comment by nzakas** [comment]
Thanks for the input. As [stated in the README](https://github.com/eslint/eslint#stylistic-rule-updates), we actually aren't making any further updates to rules that enforce stylistic preferences. If you have a strong preference for such a change, you can always copy the existing rule into a custom rule and make it behave the way you want.
